### ISSUE TITLE
[Question] File '.../node_modules/intl-locale-textinfo-polyfill/index.d.ts' is not a module.ts (2306)

### ISSUE THREAD

**Issue by gustaveWPM** [brettz9/intl-locale-textinfo-polyfill#1]
Hello,
I've installed the lib like this:

```bash
yarn add intl-locale-textinfo-polyfill
```

And then, I tried to import it like this:
```ts
// dummy-ts-file.ts
import Locale from 'intl-locale-textinfo-polyfill';
```

Then, I have an error:
> File '.../node_modules/intl-locale-textinfo-polyfill/index.d.ts' is not a module.ts(2306)

So, what I did to trick the TS compiler is to create a `@types/intl-locale-textinfo-polyfill.d.ts` at the root of my project.

```ts
// @types/intl-locale-textinfo-polyfill.d.ts
declare module 'intl-locale-textinfo-polyfill';
```

And now, all works properly.

That's weird.
Do you have any idea why I have to create this rather inaccurate TypeScript definition file myself?

Regards.

**Comment by brettz9** [comment]
Yeah, it looks like my types were for the polyfill file instead of the ponyfill. I'm a bit busy at the moment to look to fix this properly, but it seems you've identified a workaround.

**Comment by gustaveWPM** [comment]
> Yeah, it looks like my types were for the polyfill file instead of the ponyfill. I'm a bit busy at the moment to look to fix this properly, but it seems you've identified a workaround.

In fact, I get trolled by my workflow.
When I wrote this:

```ts
import 'intl-locale-textinfo-polyfill';
```

And then saved my .ts file... then VSCode just removed the import (based on my coding style config). :/

Would be glad to have some news from you.

**Comment by brettz9** [comment]
Why does your workflow strip out imports? This is the approach for polyfills.

**Comment by gustaveWPM** [comment]
> Why does your workflow strip out imports? This is the approach for polyfills.

Sorry, I'm tired and talking non-sense... 
I had refactored a lot my code, and I thought my editor had stripped the import, but in fact it was moved in another file. u_u

(Sorry again!)
The rest of the outcome is still valid, though.

**Comment by brettz9** [comment]
I've published v2.0.2. Do you want to let me know if that works for you?

**Comment by gustaveWPM** [comment]
> I've published v2.0.2. Do you want to let me know if that works for you?

I'll take a look right now

**Comment by brettz9** [comment]
Note that to get the polyfill version, you have to require it directly, i.e., `import 'intl-locale-textinfo-polyfill/lib/polyfill.js'`.

**Comment by gustaveWPM** [comment]
Well...
When I import the polyfill like this:

```ts
import Locale from 'intl-locale-textinfo-polyfill';
```

I get an error:

> File '/home/stan/Delivery/dashboard_rtm/node_modules/intl-locale-textinfo-polyfill/index.d.ts' is not a module.ts(2306)

When I import it like this:

```ts
import 'intl-locale-textinfo-polyfill/lib/polyfill.js'

// * ...
const { direction: dir } = new Intl.Locale(locale).textInfo;
```

I get an error:

> Property 'textInfo' does not exist on type 'Locale'.ts(2339)

When I import like this:

```ts
import 'intl-locale-textinfo-polyfill';

// * ...
const { direction: dir } = new Intl.Locale(locale).textInfo;
```

I get an error:

> Property 'textInfo' does not exist on type 'Locale'.ts(2339)

Hummm...
Idk what I'm doing wrong. :(

Please, could you give a look with the following project?
https://github.com/Tirraa/dashboard_rtm

The TS file which is raising the error:
https://github.com/Tirraa/dashboard_rtm/blob/main/src/components/layouts/base/DocumentRoot.tsx

Workaround file:
https://github.com/Tirraa/dashboard_rtm/blob/main/%40types/intl-locale-textinfo-polyfill.d.ts

**Comment by brettz9** [comment]
I've published v2.1.0 which might fix the polyfill error. I'm not sure what the ponyfill error is about. It sounds like you may need to restart vscode.

I may be doing something wrong here myself, so don't think it is necessarily something in your project.

**Comment by gustaveWPM** [comment]
Hmmmm... In fact, I can import `Locale` like this:

```ts
import Locale from 'lib/Locale';
```

That's curious.
I think 'lib/Locale' shouldn't be exposed like this?

---

Secondly, maybe I have an idea for the error:
> File '.../node_modules/intl-locale-textinfo-polyfill/index.d.ts' is not a module.ts(2306)

I'm a TypeScript newby.
So I may write something wrong, but...

Maybe we should turn:

```ts
// intl-locale-textinfo-polyfill/lib/polyfill.d.ts

declare namespace Intl {
  class Locale {
    constructor(locale: string);
    get textInfo(): {
        direction: "ltr"|"rtl";
    };
  }
}
```

To something like:
```ts
// intl-locale-textinfo-polyfill/lib/polyfill.d.ts

declare module 'intl-locale-textinfo-polyfill' {
  export class Locale {
    constructor(locale: string);
    get textInfo(): {
      direction: "ltr" | "rtl";
    };
  }
}
```

?

I will try to investigate on my side.

**Comment by brettz9** [comment]
Before trying that, do be sure to restart vscode after installing my latest code, and trying both the non-polyfill and polyfill versions again.

**Comment by gustaveWPM** [comment]
> Before trying that, do be sure to restart vscode after installing my latest code, and trying both the non-polyfill and polyfill versions again.

I always use the _Developer: Reload Window_ command, and try to run a `yarn build` which processes on-build type checks.

I don't know TypeScript well enough to know exactly what to do to fix this issue, sorry.
I'll try to ask to some other devs.

**Comment by gustaveWPM** [comment]
Hey @brettz9 

Could you give a look to this? 🙂 

I think the first answer talking about the "global" namespace is what we need. 

https://stackoverflow.com/questions/37640785/how-do-you-add-polyfills-to-globals-in-typescript-modules

**Comment by gustaveWPM** [comment]
Also, this should cover the "File is not a module" error:

https://stackoverflow.com/questions/32805559/typescript-es6-import-module-file-is-not-a-module-error

**Comment by brettz9** [comment]
I haven't yet fixed the polyfill, but I found the issue was apparently with `outFile` not being possible with modules. Update to the latest if you want to use the ponyfill.

**Comment by brettz9** [comment]
The polyfill is I think conflicting with the built in types for Intl.Locale.

**Comment by gustaveWPM** [comment]
Nice!
Thank you, the import of the ponyfill seems to be fixed. :)

```ts
import Locale from 'intl-locale-textinfo-polyfill';
```
Doesn't raise error anymore on my project.

**Comment by gustaveWPM** [comment]
However, I still have an error with this use case:

```ts
import 'intl-locale-textinfo-polyfill';
// * ...
const { direction: dir } = new Intl.Locale(locale).textInfo;
```

> Property 'textInfo' does not exist on type 'Locale'.ts(2339)

Personally, it doesn't bother me, as the ponyfill is enough for me. ^^

**Comment by brettz9** [comment]
Yeah, I can't figure out how to fix the polyfill. Glad to hear the ponyfill is working for you.

**Comment by gustaveWPM** [comment]
Maybe you could just drop it.
This is just a polyfill, and will be abandoned as soon as the standard will be implemented on all browsers.

Maybe you will find how to fix it later, and should just update the documentation for now?

I'm sorry to be so helpless. I really don't know how to properly fix it and how to properly override the default definition too for now. :/

**Comment by gustaveWPM** [comment]
Hello @brettz9 

I got two feedbacks.

> Currently, this library is incompatible with recent TypeScript frameworks, especially Next.js, because it is written in pure JavaScript. This means that important files used for dynamic typing and developer documentation will be missing. (Also, note that Next.js uses the 'next' modules instead of the commonjs modules).
> 
> Ideally, the author of the library should consider releasing a major version of this library in TypeScript. Once transpiled, it would be compatible with both Next.js and common JavaScript, making it more usable.

---

> Don't forget that the standard is available here, so you can write the most compliant polyfill possible.
> https://tc39.es/proposal-intl-locale-info/#sec-Intl.Locale.prototype.getTextInfo

---

Hope this helps.

**Comment by brettz9** [comment]
Thanks for the info, but that shouldn't be an issue. Although it is in pure JavaScript, it is JavaScript with JSDoc, allowing it to take advantage of types, and I was able to build the non-polyfill type declarations because of this. There is just the issue of providing appropriate types for the polyfill file. Since I have manually created that, I think it is just an issue of finding out what TypeScript would generate.

As far as the current standard, there have since been updates, but unfortunately, Chrome does not yet support it (e.g., using `getTextInfo` instead of `textInfo`), but as soon as there is some support in the browser, I would like to update at such time.
